In [1]:
import os
import tarfile 
import urllib 
DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml2/master/" 
HOUSING_PATH = os.path.join("datasets", "housing") 
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz" 
def fetch_housing_data( housing_url = HOUSING_URL, housing_path = HOUSING_PATH): 
    os.makedirs(housing_path, exist_ok = True) 
    tgz_path = os.path.join(housing_path, "housing.tgz") 
    urllib.request.urlretrieve(housing_url, tgz_path) 
    housing_tgz = tarfile.open(tgz_path)
    housing_tgz.extractall(path=housing_path)
    housing_tgz.close()    
fetch_housing_data()

import pandas as pd
def load_housing_data(housing_path=HOUSING_PATH):
    csv_path = os.path.join(housing_path, "housing.csv")
    return pd.read_csv(csv_path)
housing = load_housing_data()
#housing.head()
#housing.info()
#housing["ocean_proximity"].value_counts()
#housing.describe()
import matplotlib.pyplot as plt
housing.hist(bins=50, figsize=(20,15))
plt.show()

import numpy as np

def split_train_test(data, test_ratio):
    shuffled_indices = np.random.permutation(len(data))
    test_set_size = int(len(data) * test_ratio)
    test_indices = shuffled_indices[:test_set_size]
    train_indices = shuffled_indices[test_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

train_set, test_set = split_train_test(housing, 0.2)
#len(train_set)
#len(test_set)

from zlib import crc32

def test_set_check(identifier, test_ratio):
    return crc32(np.int64(identifier)) & 0xffffffff < test_ratio * 2**32

def split_train_test_by_id(data, test_ratio, id_column):
    ids = data[id_column]
    in_test_set = ids.apply(lambda id_: test_set_check(id_, test_ratio))
    return data.loc[~in_test_set], data.loc[in_test_set]

"""
housing_with_id = housing.reset_index() #adds an 'index' column
housing_with_id["id"] = housing["longitude"] * 1000 + housing["latitude"]
train_set, test_set = split_train_test_by_id(housing_with_id, 0.2, "index")
"""

# Does the same thing as function above with additional features like random_stat parameter allows generator seed
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(housing, test_size=0.2, random_state=42)

housing["income_cat"] = pd.cut(housing["median_income"], bins=[0., 1.5, 3.0, 4.5, 6., np.inf], labels=[1,2,3,4,5])
housing["income_cat"].hist()

from sklearn.model_selection import StratifiedShuffleSplit
split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
for train_index, test_index in split.split(housing, housing["income_cat"]):
    strat_train_set = housing.loc[train_index]
    strat_test_set = housing.loc[test_index]

strat_test_set["income_cat"].value_counts() / len(strat_test_set)

for set_ in (strat_train_set, strat_test_set):
    set_.drop("income_cat", axis=1, inplace=True)

housing = strat_train_set.copy()
housing.plot(kind="scatter", x="longitude", y="latitude", alpha=0.4, s=housing["population"]/100, 
             label="population", figsize=(10,7), c="median_house_value", cmap=plt.get_cmap("jet"), colorbar=True,)
plt.legend()
plt.tight_layout()

corr_matrix = housing.corr()
corr_matrix["median_house_value"].sort_values(ascending=False)

from pandas.plotting import scatter_matrix
attributes = ["median_house_value", "median_income", "total_rooms", "housing_median_age"]
scatter_matrix(housing[attributes], figsize=(12,8))

housing.plot(kind="scatter", x="median_income", y="median_house_value", alpha=0.1)

housing["rooms_per_household"] = housing["total_rooms"]/housing["households"]
housing["bedrooms_per_room"] = housing["total_bedrooms"]/housing["total_rooms"]
housing["population_per_household"] = housing["population"]/housing["households"]
corr_matrix = housing.corr()
corr_matrix["median_house_value"].sort_values(ascending=False)

<Figure size 2000x1500 with 9 Axes>

median_house_value          1.000000
median_income               0.687160
rooms_per_household         0.146285
total_rooms                 0.135097
housing_median_age          0.114110
households                  0.064506
total_bedrooms              0.047689
population_per_household   -0.021985
population                 -0.026920
longitude                  -0.047432
latitude                   -0.142724
bedrooms_per_room          -0.259984
Name: median_house_value, dtype: float64

In [2]:
housing = strat_train_set.drop("median_house_value", axis=1,)
housing_labels = strat_train_set["median_house_value"].copy()

In [3]:
# Data Cleaning
"""
housing.dropna( subset =[" total_bedrooms"]) # option 1: Get rid of the corresponding districts
housing.drop("total_bedrooms", axis=1) # option 2: Get rid of whole attribute
median = housing["total_bedrooms"].median() #option 3: Set the values to some value(zero, the mean, the median,ect)
"""
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy="median")

housing_num = housing.drop("ocean_proximity", axis=1)
imputer.fit(housing_num)
imputer.statistics_
housing_num.median().values

array([-118.51  ,   34.26  ,   29.    , 2119.5   ,  433.    , 1164.    ,
        408.    ,    3.5409])

In [4]:
X = imputer.transform(housing_num)
housing_tr = pd.DataFrame(X, columns=housing_num.columns, index=housing_num.index)

In [5]:
housing_cat = housing[["ocean_proximity"]]
#housing_cat.head(10)
from sklearn.preprocessing import OrdinalEncoder
ordinal_encoder = OrdinalEncoder()
housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat)
housing_cat_encoded[:10]
ordinal_encoder.categories_

[array(['<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN'],
       dtype=object)]

In [6]:
from sklearn.preprocessing import OneHotEncoder
cat_encoder = OneHotEncoder()
housing_cat_1hot = cat_encoder.fit_transform(housing_cat)
#housing_cat_1hot.toarray()
housing_cat_1hot

<16512x5 sparse matrix of type '<class 'numpy.float64'>'
	with 16512 stored elements in Compressed Sparse Row format>

In [7]:
# Custom Transformer; Small transformer class that adds the combined attributes discussed earlier
from sklearn.base import BaseEstimator, TransformerMixin
rooms_ix, bedrooms_ix, population_ix, households_ix = 3, 4, 5, 6
class CombinedAttributesAdder(BaseEstimator, TransformerMixin): 
    def __init__(self, add_bedrooms_per_room = True):  # no *args or ** kargs 
        self.add_bedrooms_per_room = add_bedrooms_per_room 
    def fit( self, X, y = None): return self # nothing else to do 
    def transform( self, X):
        rooms_per_household = X[:, rooms_ix] / X[:, households_ix] 
        population_per_household = X[:, population_ix] / X[:, households_ix] 
        # add_bedrooms_per_room is a hyperparameter controlling whether we add this attribute or not
        if self.add_bedrooms_per_room: 
            bedrooms_per_room = X[:, bedrooms_ix] / X[:, rooms_ix] 
            return np.c_[ X, rooms_per_household, population_per_household, bedrooms_per_room] 
        else: 
            return np.c_[ X, rooms_per_household, population_per_household] 
attr_adder = CombinedAttributesAdder( add_bedrooms_per_room = False)
housing_extra_attribs = attr_adder.transform(housing.values) 

In [8]:
# Transformation pipelines
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
num_pipeline = Pipeline([
        (' imputer', SimpleImputer( strategy ="median")), 
        (' attribs_adder', CombinedAttributesAdder()), 
        (' std_scaler', StandardScaler()), ]) 
housing_num_tr = num_pipeline.fit_transform(housing_num)

In [9]:
from sklearn.compose import ColumnTransformer 
num_attribs = list(housing_num) 
cat_attribs = ["ocean_proximity"] 
full_pipeline = ColumnTransformer([("num", num_pipeline, num_attribs), ("cat", OneHotEncoder(), cat_attribs),]) 
housing_prepared = full_pipeline.fit_transform(housing)

In [10]:
# Select and train on the data set
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(housing_prepared, housing_labels)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [11]:
# trying the training on a few instances of the training set
some_data = housing.iloc[:5]
some_labels = housing_labels.iloc[:5]
some_data_prepared = full_pipeline.transform(some_data)
print("Predictions:", lin_reg.predict(some_data_prepared))
print("Labels:", list(some_labels))

Predictions: [210644.60459286 317768.80697211 210956.43331178  59218.98886849
 189747.55849879]
Labels: [286600.0, 340600.0, 196900.0, 46300.0, 254500.0]


In [12]:
# measuring the regression model's RMSE on the whole training set
from sklearn.metrics import mean_squared_error
housing_predictions = lin_reg.predict(housing_prepared)
lin_mse = mean_squared_error(housing_labels, housing_predictions)
lin_rmse = np.sqrt(lin_mse)
lin_rmse # how much off the predictions are

68628.19819848923

In [13]:
# DecisionTreeRegressor model which is more powerful
from sklearn.tree import DecisionTreeRegressor
tree_reg = DecisionTreeRegressor()
tree_reg.fit(housing_prepared, housing_labels)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')

In [14]:
housing_predictions = tree_reg.predict(housing_prepared)
tree_mse = mean_squared_error(housing_labels, housing_predictions)
tree_rmse = np.sqrt(tree_mse)
tree_rmse # how much off the predictions are

0.0

In [15]:
# better evaluation with cross evaluation
from sklearn.model_selection import cross_val_score
scores = cross_val_score(tree_reg, housing_prepared, housing_labels, scoring ="neg_mean_squared_error", cv = 10) 
tree_rmse_scores = np.sqrt(-scores) 

In [16]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation:", scores.std()) 

In [17]:
display_scores(tree_rmse_scores)

Scores: [69238.8754295  67086.16158399 72188.15345757 69696.62749101
 69787.90846948 74797.16057015 71747.89629623 71114.01348053
 76177.21842724 71714.6589164 ]
Mean: 71354.86741221012
Standard deviation: 2530.7141454633347


In [18]:
lin_scores = cross_val_score(lin_reg, housing_prepared, housing_labels, scoring ="neg_mean_squared_error", cv = 10)
lin_rmse_scores = np.sqrt(-lin_scores)
display_scores(lin_rmse_scores)

Scores: [66782.73843989 66960.118071   70347.95244419 74739.57052552
 68031.13388938 71193.84183426 64969.63056405 68281.61137997
 71552.91566558 67665.10082067]
Mean: 69052.46136345083
Standard deviation: 2731.6740017983493


In [19]:
# Random Forest Regressor model
from sklearn.ensemble import RandomForestRegressor
forest_reg = RandomForestRegressor()
forest_reg.fit(housing_prepared, housing_labels)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=10,
                      n_jobs=None, oob_score=False, random_state=None,
                      verbose=0, warm_start=False)

In [20]:
housing_predictions = forest_reg.predict(housing_prepared)
forest_mse = mean_squared_error(housing_labels, housing_predictions)
forest_rmse = np.sqrt(forest_mse)
forest_rmse

forest_scores = cross_val_score(forest_reg, housing_prepared, housing_labels, scoring="neg_mean_squared_error", cv=10)
forest_rmse_scores = np.sqrt(-forest_scores)
display_scores(forest_rmse_scores)
# you should save your model, but not going to do it rn. page 75 in book.

Scores: [52043.17116335 51423.90316831 53585.59107417 53939.80126411
 52738.59112243 56332.3000374  50857.06136825 50341.80047545
 55445.13121956 52159.15242388]
Mean: 52886.65033169356
Standard deviation: 1844.2810158446498


In [21]:
from sklearn.model_selection import GridSearchCV

param_grid = [
    # try 12 (3×4) combinations of hyperparameters
    {'n_estimators': [3, 10, 30], 'max_features': [2, 4, 6, 8]},
    # then try 6 (2×3) combinations with bootstrap set as False
    {'bootstrap': [False], 'n_estimators': [3, 10], 'max_features': [2, 3, 4]},
  ]

forest_reg = RandomForestRegressor(random_state=42)
# train across 5 folds, that's a total of (12+6)*5=90 rounds of training 
grid_search = GridSearchCV(forest_reg, param_grid, cv=5,
                           scoring='neg_mean_squared_error',
                           return_train_score=True)
grid_search.fit(housing_prepared, housing_labels)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=RandomForestRegressor(bootstrap=True, criterion='mse',
                                             max_depth=None,
                                             max_features='auto',
                                             max_leaf_nodes=None,
                                             min_impurity_decrease=0.0,
                                             min_impurity_split=None,
                                             min_samples_leaf=1,
                                             min_samples_split=2,
                                             min_weight_fraction_leaf=0.0,
                                             n_estimators='warn', n_jobs=None,
                                             oob_score=False, random_state=42,
                                             verbose=0, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid=[{'max_features': [2, 4, 6, 8],
  

In [22]:
grid_search.best_params_

{'max_features': 8, 'n_estimators': 30}

In [23]:
grid_search.best_estimator_

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features=8, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=30,
                      n_jobs=None, oob_score=False, random_state=42, verbose=0,
                      warm_start=False)

In [24]:
cvres = grid_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)    

63669.05791727153 {'max_features': 2, 'n_estimators': 3}
55627.16171305252 {'max_features': 2, 'n_estimators': 10}
53384.57867637289 {'max_features': 2, 'n_estimators': 30}
60965.99185930139 {'max_features': 4, 'n_estimators': 3}
52740.98248528835 {'max_features': 4, 'n_estimators': 10}
50377.344409590376 {'max_features': 4, 'n_estimators': 30}
58663.84733372485 {'max_features': 6, 'n_estimators': 3}
52006.15355973719 {'max_features': 6, 'n_estimators': 10}
50146.465964159885 {'max_features': 6, 'n_estimators': 30}
57869.25504027614 {'max_features': 8, 'n_estimators': 3}
51711.09443660957 {'max_features': 8, 'n_estimators': 10}
49682.25345942335 {'max_features': 8, 'n_estimators': 30}
62895.088889905004 {'bootstrap': False, 'max_features': 2, 'n_estimators': 3}
54658.14484390074 {'bootstrap': False, 'max_features': 2, 'n_estimators': 10}
59470.399594730654 {'bootstrap': False, 'max_features': 3, 'n_estimators': 3}
52725.01091081235 {'bootstrap': False, 'max_features': 3, 'n_estimators'

In [25]:
final_model = grid_search.best_estimator_

X_test = strat_test_set.drop("median_house_value", axis=1)
y_test = strat_test_set["median_house_value"].copy()

X_test_prepared = full_pipeline.transform(X_test)
final_predictions = final_model.predict(X_test_prepared)

final_mse = mean_squared_error(y_test, final_predictions)
final_rmse = np.sqrt(final_mse)
final_rmse

47730.22690385927

In [26]:
from scipy import stats

confidence = 0.95
squared_errors = (final_predictions - y_test) ** 2
np.sqrt(stats.t.interval(confidence, len(squared_errors) - 1,
                         loc=squared_errors.mean(),
                         scale=stats.sem(squared_errors)))

array([45685.10470776, 49691.25001878])

**1.** Try a Support Vector Machine regressor (`sklearn.svm.SVR`) with various hyperparameters, such as `kernel = "linear"` (with various values for the C hyperparameter) or `kernel = "rbf"` (with various values for the C and gamma hyperparameters). Don’t worry about what these hyperparameters mean for now. How does the best SVR predictor perform?

In [28]:
# Support Vector Machine Regressor with Grid search
from sklearn.svm import SVR
param_grid = [
    # try 5 (1×5) combinations of hyperparameters
    {'kernel': ['linear'], 'C': [10.0,50.0,100.0,500.0,1000.0]},
    # then try 25 (1x5×5) combinations of hyperparameters
    {'kernel': ['rbf'], 'C': [1,10.0,50.0,100.0,500.0], 'gamma': [0.01, 0.03, 0.1, 0.3, 1.0, 3.0]},
  ]

support_vector_reg = SVR()
# train across 5 folds, that's a total of (5+25)*5=150 rounds of training 
grid_search = GridSearchCV(support_vector_reg, param_grid, cv=5,
                           scoring='neg_mean_squared_error')
grid_search.fit(housing_prepared, housing_labels)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3,
                           epsilon=0.1, gamma='auto_deprecated', kernel='rbf',
                           max_iter=-1, shrinking=True, tol=0.001,
                           verbose=False),
             iid='warn', n_jobs=None,
             param_grid=[{'C': [10.0, 50.0, 100.0, 500.0, 1000.0],
                          'kernel': ['linear']},
                         {'C': [1, 10.0, 50.0, 100.0, 500.0],
                          'gamma': [0.01, 0.03, 0.1, 0.3, 1.0, 3.0],
                          'kernel': ['rbf']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='neg_mean_squared_error', verbose=0)

In [44]:
mse = -(grid_search.best_score_)
rmse = np.sqrt(mse)
print("Best SVR predictor's performance: ", rmse, '\n')
cvres = grid_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

Best SVR predictor's performance:  70445.41077712944 

84654.0893002138 {'C': 10.0, 'kernel': 'linear'}
73235.0217516481 {'C': 50.0, 'kernel': 'linear'}
71603.12196479437 {'C': 100.0, 'kernel': 'linear'}
70538.40564891648 {'C': 500.0, 'kernel': 'linear'}
70445.41077712944 {'C': 1000.0, 'kernel': 'linear'}
118810.1205690511 {'C': 1, 'gamma': 0.01, 'kernel': 'rbf'}
118692.80532644158 {'C': 1, 'gamma': 0.03, 'kernel': 'rbf'}
118637.95452846108 {'C': 1, 'gamma': 0.1, 'kernel': 'rbf'}
118768.15436248068 {'C': 1, 'gamma': 0.3, 'kernel': 'rbf'}
118886.40095734302 {'C': 1, 'gamma': 1.0, 'kernel': 'rbf'}
118916.0277544163 {'C': 1, 'gamma': 3.0, 'kernel': 'rbf'}
117855.86860610843 {'C': 10.0, 'gamma': 0.01, 'kernel': 'rbf'}
116721.30926754372 {'C': 10.0, 'gamma': 0.03, 'kernel': 'rbf'}
116179.73058899572 {'C': 10.0, 'gamma': 0.1, 'kernel': 'rbf'}
117392.89093611801 {'C': 10.0, 'gamma': 0.3, 'kernel': 'rbf'}
118595.3378452412 {'C': 10.0, 'gamma': 1.0, 'kernel': 'rbf'}
118880.72530621996 {'C': 10.

The final RMSE value was 47730.22690385927 so in comparison this is much worse

**2.** Try replacing `GridSearchCV` with `RandomizedSearchCV`.

In [60]:
# Support Vector Machine Regressor with Random Search
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import expon, reciprocal

param_dist = {'kernel': ['linear', 'rbf'], 'C': reciprocal(20,200000), 'gamma': expon(scale=1.0)}

support_vector_reg = SVR()
random_search = RandomizedSearchCV(support_vector_reg, param_distributions=param_dist, cv=5, n_iter=50,
                           scoring='neg_mean_squared_error', random_state=42)
random_search.fit(housing_prepared, housing_labels)

RandomizedSearchCV(cv=5, error_score='raise-deprecating',
                   estimator=SVR(C=1.0, cache_size=200, coef0=0.0, degree=3,
                                 epsilon=0.1, gamma='auto_deprecated',
                                 kernel='rbf', max_iter=-1, shrinking=True,
                                 tol=0.001, verbose=False),
                   iid='warn', n_iter=50, n_jobs=None,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x0000016DAA33FA88>,
                                        'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x0000016DAA33FD08>,
                                        'kernel': ['linear', 'rbf']},
                   pre_dispatch='2*n_jobs', random_state=42, refit=True,
                   return_train_score=False, scoring='neg_mean_squared_error',
                   verbose=0)

In [61]:
mse = -(random_search.best_score_)
rmse = np.sqrt(mse)
print("Best SVR predictor's performance: ", rmse, '\n')
cvres = random_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

Best SVR predictor's performance:  54767.99053704408 

70487.73975752539 {'C': 629.782329591372, 'gamma': 3.010121430917521, 'kernel': 'linear'}
65151.98496543675 {'C': 26290.206464300216, 'gamma': 0.9084469696321253, 'kernel': 'rbf'}
100293.08900577638 {'C': 84.14107900575871, 'gamma': 0.059838768608680676, 'kernel': 'rbf'}
70574.11375177321 {'C': 432.37884813148855, 'gamma': 0.15416196746656105, 'kernel': 'linear'}
118838.76739914888 {'C': 24.17508294611391, 'gamma': 3.503557475158312, 'kernel': 'rbf'}
70029.23360936904 {'C': 113564.03940586245, 'gamma': 0.0007790692366582295, 'kernel': 'rbf'}
107662.52814484686 {'C': 108.30488238805073, 'gamma': 0.3627537294604771, 'kernel': 'rbf'}
77562.95173075002 {'C': 21.344953672647435, 'gamma': 0.023332523598323388, 'kernel': 'linear'}
62578.76510141139 {'C': 5603.270317432516, 'gamma': 0.15023452872733867, 'kernel': 'rbf'}
54767.99053704408 {'C': 157055.10989448498, 'gamma': 0.26497040005002437, 'kernel': 'rbf'}
70364.62759201857 {'C': 27652.

The final RMSE value was 47730.22690385927 so in comparison this is much worse

**3.** Try adding a transformer in the preparation pipeline to select only the most important attributes. 